File: Java/DLinkedList.java
import java.util.NoSuchElementException;

/**
 * Created by Chris Medina on 5/16/2014.
 * <p>Purpose:  To create a Doubly-linked, double-ended Linked List that allows for insertion and deletion at both
 * ends,
 * with forwards and backwards traversal.</p>
 */
public class DLinkedList<E> {

    private Node<E> first, last;
    private int size;

    /**
     * Create an empty Linked List
     */
    public DLinkedList () {
        first = null;
        last = null;
        size = 0;
    }

    /**
     * Determine if List is empty
     *
     * @return true, if no elements are present
     */
    public boolean isEmpty () {
        return (first == null);
    }

    public String toString () {
        StringBuilder sb = new StringBuilder("[");
        for (Node<E> f = first; f != null; f = f.next) {
            if (f.equals(last))
                sb.append(f);
            else
                sb.append(f + ", ");
        }
        sb.append("]");
        return sb.toString();
    }

    /**
     * Empty the list
     */
    public void clearAll () {
        first = null;
        last = null;
        size = 0;
    }

    /**
     * Insert a Node at the front of the list
     *
     * @param val the value to insert
     */
    public void insertFirst (E item) {
        Node<E> node = new Node<>(null, item, first);
        if (isEmpty())
            last = node;
        else {
            first.previous = node;
        }
        size++;
        first = node;
    }

    /**
     * Insert a Node to the end of the list
     *
     * @param val the value to insert
     */
    public void insertLast (E item) {
        Node<E> node = new Node<>(last, item, null);
        if (isEmpty())
            first = node;
        else {
            last.next = node;
        }
        size++;
        last = node;
    }


    /**
     * Insert a Node in front of another (right of)
     *
     * @param find the value to insert after
     * @param val  the value to insert
     */
    public void insertAfter (E find, E item) {
        Node<E> curr = first;
        while (curr.val != find) {
            curr = curr.next;
            if (curr == null)
                return;
        }

        Node<E> node = new Node<>(curr, item, curr.next);
        if (curr == first) {
            if (first.next == null) {
                last = node;
            } else
                first.next.previous = node;
            first.next = node;
        } else if (curr == last) {
            last.next = node;
            last = node;
        } else {
            curr.next.previous = node;
            curr.next = node;
        }
        size++;
    }

    /**
     * Remove this value from the Linked List
     *
     * @param val the value to remove
     * @return the Node removed
     * @throws java.util.NoSuchElementException if value not found
     */
    public Node<E> remove (E find) {
        Node<E> curr = first;
        while (curr.val != find) {
            curr = curr.next;
            if (curr == null)
                throw new NoSuchElementException("Value not found in List");
        }

        if (curr == first) {
            if (first.next == null) {
                Node<E> tempFirst = first;
                first = null;
                last = null;
                return tempFirst;
            } else
                first.next.previous = null;
            first = first.next;
        } else if (curr == last) {
            last.previous.next = null;
            last = last.previous;
        } else {
            curr.next.previous = curr.previous;
            curr.previous.next = curr.next;
        }

        size--;
        return curr;
    }

    /**
     * Remove the first Node from the List
     *
     * @return the head node
     */
    public Node<E> removeFirst () {
        return remove(first.val);
    }

    /**
     * Remove the last Node from the List
     *
     * @return the last node
     */
    public Node<E> removeLast () {
        return remove(last.val);
    }


    /**
     * Remove every nth Node from this List
     *
     * @param n The number of steps we take until we remove the current Node
     */
    public void removeEveryNthNode (int n) {
        int ptr = 1;
        for (Node<E> f = first; f != null; f = f.next) {
            if (ptr == n) {
                remove(f.val);
                ptr = 1;
            } else
                ptr++;
        }
    }

    /**
     * Return a Node based on its index within the list (recursive)
     *
     * @param index the index of the element to retireve
     * @return the Node at this index
     */
    public Node<E> get (int index) {
        if (index == 0)
            return first;
        else
            return get(index - 1).next;
    }

    /**
     * Find this value within the linked list (recursive)
     *
     * @param find the value to find
     * @return true, if found
     */
    public boolean contains (Node<E> start, E find) {
        if (start != null) {
            if (start.val == find)
                return true;
            else
                return contains(start.next, find);
        } else
            return false;
    }

    /**
     * Find this value within the linked list
     *
     * @param find the value to find
     * @return true, if found
     */
    public boolean contains (E find) {
        if (isEmpty())
            return false;
        Node<E> curr = first;
        while (curr.val != find) {
            if (curr.equals(last))
                return false;
            curr = curr.next;
        }
        return true;
    }

    /**
     * Determine if List contains Duplicates
     * O(n^2)
     */
    public boolean containsDuplicates () {
        for (Node<E> prev = first; prev != null; prev = prev.next) {
            for (Node<E> curr = prev.next; curr != null; curr = curr.next) {
                if (prev.equals(curr)) ;
                return true;
            }
        }
        return false;
    }

    /**
     * Remove duplicates from the current list
     * O(n^2)
     */
    public void removeDuplicates () {
        for (Node<E> prev = first; prev != null; prev = prev.next) {
            for (Node<E> curr = prev.next; curr != null; curr = curr.next) {
                if (prev.equals(curr))
                    remove(prev.val);
            }
        }
    }

    /**
     * Find the nth-To-last element in the Linked List
     *
     * @param n the n-th last element to find
     * @return the nth-To-Last element in list
     */
    public Node<E> nthLast (int n) {
        Node<E> curr = first, prev = first;
        while (n-- > 0)
            curr = curr.next;
        while (curr != null) {
            prev = prev.next;
            curr = curr.next;
        }
        return prev;
    }

    /**
     * Find the middle Node in the current List
     *
     * @return the middle Node
     */
    public Node<E> mid () {
        Node<E> curr = first, prev = first;
        int ptr = 1;
        while (curr != null) {
            if (ptr % 2 == 0)
                prev = prev.next;
            curr = curr.next;
            ptr++;
        }
        return prev;
    }

    /**
     * The dual-way (forwards/ backwards) traversal iterator for this List
     *
     * @return an iterator for the List
     */
    public ListIterating<E> listIterator () {
        return new ListIterator();
    }

    /**
     * Class to represent the individual Nodes in the Linked List
     *
     * @param <E> generic form
     */
    private static class Node<E> {
        E val;
        Node<E> previous, next;


        private Node (Node<E> previous, E val, Node<E> next) {
            this.previous = previous;
            this.val = val;
            this.next = next;
        }

        @Override
        public String toString () {
            return String.format("%s", val);
        }

        @Override
        public boolean equals (Object obj) {
            if (obj == this)
                return true;
            if (obj == null || getClass() != obj.getClass())
                return false;

            Node<E> temp = (Node<E>) obj;
            return (this.val == temp.val);
        }

        @Override
        public int hashCode () {
            int result = 17;
            result = result + (previous == null ? 0 : previous.hashCode());
            result = result + (val == null ? 0 : val.hashCode());
            result = result + (next == null ? 0 : next.hashCode());

            return super.hashCode();
        }
    }

    /**
     * Class that provides for dual-way traversal of List
     */
    private class ListIterator implements ListIterating<E> {

        Node<E> curr, prev, next;

        private ListIterator () {
            prev = null;
            curr = null;
            next = first;
        }

        @Override
        public boolean hasPrevious () {
            return (prev != null);
        }

        @Override
        public E previous () {
            next = curr;
            curr = prev;
            prev = prev.previous;
            return curr.val;
        }

        @Override
        public boolean hasNext () {
            return (next != null);
        }

        @Override
        public E next () {
            prev = curr;
            curr = next;
            next = next.next;
            return curr.val;
        }

        @Override
        public void reset () {
            prev = null;
            curr = null;
            next = first;
        }

    }
}

File: C-Sharp/DLinkedList.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace OOP
{
    /// <summary>
    /// DLinkedList represents a double-ended, doubly-linked List data structure
    /// Date: 5-26-2014
    /// </summary>
    /// <typeparam name="E">Generic, type-safe</typeparam>
    public class DLinkedList<E>
    {
        public Node<E> First { get; private set; }
        public Node<E> Last { get; private set; }
        public int Size { get; private set; }

        /// <summary>
        /// Construct an empty Linked List
        /// </summary>
        public DLinkedList () {
            First = Last = null;
            Size = 0;
        }

        public override string ToString () {
            StringBuilder sb = new StringBuilder("[");

            for (Node<E> f = First; f != null; f = f.Next) {
                if (f.Next == null)
                    sb.Append(f);
                else
                    sb.Append(f + ", ");
            }
            sb.Append("]");
            return sb.ToString();
        }
        /// <summary>
        /// Determine if this List is empty
        /// </summary>
        /// <returns>true, if there are no items in the list</returns>
        public bool IsEmpty () {
            return (First == null);
        }

        /// <summary>
        /// Insert an item to the front of the List
        /// </summary>
        /// <param name="item"></param>
        public void InsertFirst (E item) {
            Node<E> node = new Node<E>(null, item, First);
            if (IsEmpty())
                Last = node;
            else
                First.Previous = node;
            First = node;
            Size++;
        }

        /// <summary>
        /// Insert an item to the end
        /// </summary>
        /// <param name="item"></param>
        public void InsertLast (E item) {
            Node<E> node = new Node<E>(Last, item, null);
            if (IsEmpty())
                First = node;
            else
                Last.Next = node;
            Last = node;
            Size++;
        }

        /// <summary>
        /// Insert a value immediately after another specific value
        /// </summary>
        /// <param name="find">The item to find to insert after</param>
        /// <param name="item">The actual item to insert</param>
        public void InsertAfter (E find, E item) {
            Node<E> Temp = Search(find);
            Node<E> Node = new Node<E>(Temp, item, Temp.Next);

            if (Temp == First) {
                if (Temp.Next == null)
                    Last = Node;
                else {
                    First.Next.Previous = Node;
                }
                First.Next = Node;
            }
            else if (Temp == Last) {
                Last.Next = Node;
                Last = Node;
            }
            else {
                Temp.Next.Previous = Node;
                Temp.Next = Node;
            }

            Size++;
        }

        /// <summary>
        /// Search for a specific value and return its encapsulating type(Node which stores this E-type value)
        /// </summary>
        /// <param name="find">The value to find</param>
        /// <returns>The Node that holds this value</returns>
        public Node<E> Search (E find) {
            Node<E> f = First;
            while (!f.Val.Equals(find)) {
                f = f.Next;
                if (f == null)
                    throw new ArgumentException("Value not found");
            }
            return f;
        }

        /// <summary>
        /// Remove a value from the List
        /// </summary>
        /// <param name="find">The value to remove</param>
        /// <returns>The removed value if found, or throws ArgumentException if not found</returns>
        public E Remove (E find) {
            if (!Contains(find))
                throw new ArgumentException("Value not found");
            Node<E> f = First;
            while (!f.Val.Equals(find)) {
                f = f.Next;
            }

            if (f == First) {
                if (f.Next == null) {
                    First = null;
                    Last = null;
                    return f.Val;
                }
                else {
                    First.Next.Previous = null;
                }
                First = First.Next;
            }
            else if (f == Last) {
                Last.Previous.Next = null;
                Last = Last.Previous;
            }
            else {
                f.Previous.Next = f.Next;
                f.Next.Previous = f.Previous;
            }

            Size--;
            return f.Val;
        }

        /// <summary>
        /// Remove the first value in the List
        /// </summary>
        /// <returns>The value at the front</returns>
        public E RemoveFirst () {
            return Remove(First.Val);
        }


        /// <summary>
        /// Remove the last value in the List
        /// </summary>
        /// <returns>The value at the end</returns>
        public E RemoveLast () {
            return Remove(Last.Val);
        }

        /// <summary>
        /// Determine if this value is present in the List
        /// </summary>
        /// <param name="find">The value to search for</param>
        /// <returns>true, if found</returns>
        public bool Contains (E find) {
            Node<E> f = First;
            while (!f.Val.Equals(find)) {
                f = f.Next;
                if (f == null)
                    return false;
            }
            return true;
        }
    }

    /// <summary>
    /// Encapsulating data-store class which holds the information about each entry in the Linked List
    /// </summary>
    /// <typeparam name="E">generic, type-safe</typeparam>
    public sealed class Node<E>
    {
        public E Val { get; set; }
        public Node<E> Next { get; set; }
        public Node<E> Previous { get; internal set; }

        public Node (Node<E> prev, E val, Node<E> next) {
            this.Val = val;
            this.Previous = prev;
            this.Next = next;
        }

        public override string ToString () {
            return String.Format("{0}", this.Val);
        }
    }

}
